internal/resource/obclusteroperation/obclusteroperation_task.go

```diff
@@ -0,0 +1,248 @@
+			obcluster.Spec.MonitorTemplate = m.Resource.Spec.RestartOBServers.AddingMonitor
+		}
+		if m.Resource.Spec.RestartOBServers.AddingBackupVolume != nil {
+			obcluster.Spec.BackupVolume = m.Resource.Spec.RestartOBServers.AddingBackupVolume
+		}
+	case strings.ToLower(string(constants.ClusterOpTypeUpgrade)):
+		if m.Resource.Spec.Upgrade == nil {
+			return errors.New("Upgrade is empty")
+		}
+		if m.Resource.Spec.Upgrade.Image == "" {
+			return errors.New("Upgrading image is empty")
+		}
+		obcluster.Spec.OBServerTemplate.Image = m.Resource.Spec.Upgrade.Image
+	case strings.ToLower(string(constants.ClusterOpTypeExpandStorageSize)):
+		if m.Resource.Spec.ExpandStorageSize == nil {
+			return errors.New("ModifyStorageSize is empty")
+		}
+		mutation := m.Resource.Spec.ExpandStorageSize
+		if mutation.DataStorage != nil {
+			obcluster.Spec.OBServerTemplate.Storage.DataStorage.Size = *mutation.DataStorage
+		}
+		if mutation.LogStorage != nil {
+			obcluster.Spec.OBServerTemplate.Storage.LogStorage.Size = *mutation.LogStorage
+		}
+		if mutation.RedoLogStorage != nil {
+			obcluster.Spec.OBServerTemplate.Storage.RedoLogStorage.Size = *mutation.RedoLogStorage
+		}
+	case strings.ToLower(string(constants.ClusterOpTypeModifyStorageClass)):
+		if m.Resource.Spec.ModifyStorageClass == nil {
+			return errors.New("ModifyStorageClass is empty")
+		}
+		mutation := m.Resource.Spec.ModifyStorageClass
+		if mutation.DataStorage != "" {
+			obcluster.Spec.OBServerTemplate.Storage.DataStorage.StorageClass = mutation.DataStorage
+		}
+		if mutation.LogStorage != "" {
+			obcluster.Spec.OBServerTemplate.Storage.LogStorage.StorageClass = mutation.LogStorage
+		}
+		if mutation.RedoLogStorage != "" {
+			obcluster.Spec.OBServerTemplate.Storage.RedoLogStorage.StorageClass = mutation.RedoLogStorage
+		}
+	case strings.ToLower(string(constants.ClusterOpTypeSetParameters)):
+		if m.Resource.Spec.SetParameters == nil {
+			return errors.New("setParameters is empty")
+		}
+		newParamMap := make(map[string]string)
+		for _, v := range m.Resource.Spec.SetParameters {
+			newParamMap[v.Name] = v.Value
+		}
+		existingMap := make(map[string]struct{})
+		for i, v := range obcluster.Spec.Parameters {
+			if val, ok := newParamMap[v.Name]; ok {
+				obcluster.Spec.Parameters[i].Value = val
+			}
+			existingMap[v.Name] = struct{}{}
+		}
+		for k, v := range newParamMap {
+			if _, ok := existingMap[k]; !ok {
+				obcluster.Spec.Parameters = append(obcluster.Spec.Parameters, apitypes.Parameter{Name: k, Value: v})
+			}
+		}
+	}
+	if m.Resource.Spec.Force {
+		obcluster.Status.Status = clusterstatus.Running
+		obcluster.Status.OperationContext = nil
+	}
+	err = m.Client.Patch(m.Ctx, obcluster, client.MergeFrom(origin))
+	if err != nil {
+		m.Logger.Error(err, "Failed to patch obcluster")
+		return err
+	}
+
+	err = m.waitForOBClusterToBeStatus(obcfg.GetConfig().Time.DefaultStateWaitTimeout, func(status string) bool {
```

Comment:
what if the original status of obcluster is not running and this is not a force operation